feat: Initial version

wolfy1339 opened the pull request.

wolfy1339 committed:
docs(README): remove WIP note

wolfy1339 committed:
build(package): initial version

wolfy1339 committed:
build(package): lock file

wolfy1339 committed:
build(gitignore): node_modules

wolfy1339 committed:
feat: initial version

wolfy1339 committed:
chore(lint): add eslint config

wolfy1339 committed:
feat(script): initial behaviour

wolfy1339 committed:
feat: narrow implementation

gr2m commented on script.js:
```diff
@@ -0,0 +1,262 @@
+        content: prettier.format(stringify(yamlDocument), { parser: 'yaml' })
+      });
+
+      octokit.log.info('Issue created for \'%s\': %s', repo, commit.html_url);
```
Should this be 
```suggestion
      octokit.log.info('Commit created for \'%s\': %s', repo, commit.html_url);
```

gr2m reviewed (commented):
It might be simpler to use https://github.com/gr2m/octokit-plugin-create-pull-request?

wolfy1339 committed:
refactor: rewrite script to use `octokit-plugin-create-pull-request`

wolfy1339 committed:
chore(eslintrc): set ecmaVersion to 2021

wolfy1339 commented:
> It might be simpler to use https://github.com/gr2m/octokit-plugin-create-pull-request?

Indeed it is much simpler!

wolfy1339 committed:
fix: correct typo in `Array#includes()` function call

wolfy1339 committed:
fix: pass through file object

In order to get the file name

wolfy1339 committed:
fix: little fixes

- Decode the content before using it
- It's Array#push() not Array#append()
- It's BREAKING CHANGE, why can't the plural also be used

wolfy1339 commented:
I only seem to have one problem with the script.

The endpoint `/repos/{owner}/{repository}/git/trees` returns a 404 when trying to do a `POST` request (via `octokit-plugin-create-pull-request`)

Here's an extract of the log for the failed request
```json
{
    "repository": {
        "id": 366884413,
        "owner": {
            "login": "wolfy1339",
            "id": 4595477,
            "node_id": "MDQ6VXNlcjQ1OTU0Nzc=",
            "avatar_url": "https://avatars.githubusercontent.com/u/4595477?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/wolfy1339",
            "html_url": "https://github.com/wolfy1339",
            "followers_url": "https://api.github.com/users/wolfy1339/followers",
            "following_url": "https://api.github.com/users/wolfy1339/following{/other_user}",
            "gists_url": "https://api.github.com/users/wolfy1339/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/wolfy1339/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/wolfy1339/subscriptions",
            "organizations_url": "https://api.github.com/users/wolfy1339/orgs",
            "repos_url": "https://api.github.com/users/wolfy1339/repos",
            "events_url": "https://api.github.com/users/wolfy1339/events{/privacy}",
            "received_events_url": "https://api.github.com/users/wolfy1339/received_events",
            "type": "User",
            "site_admin": false
        },
        "name": "test-repository"
    },
    "msg": "HttpError: Not Found\n    at /home/user/Desktop/Programming/octoherd-remove-eol-node-versions/node_modules/@octokit/request/dist-node/index.js:86:21\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async Job.doExecute (/home/user/Desktop/Programming/octoherd-remove-eol-node-versions/node_modules/bottleneck/light.js:405:18)",
    "name": "HttpError",
    "status": 404,
    "response": {
        "url": "https://api.github.com/repos/wolfy1339/test-repository/git/trees",
        "status": 404,
        "headers": {
            "access-control-allow-origin": "*",
            "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
            "connection": "close",
            "content-encoding": "gzip",
            "content-security-policy": "default-src 'none'",
            "content-type": "application/json; charset=utf-8",
            "date": "Tue, 28 Jun 2022 16:22:38 GMT",
            "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
            "server": "GitHub.com",
            "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
            "transfer-encoding": "chunked",
            "vary": "Accept-Encoding, Accept, X-Requested-With",
            "x-accepted-oauth-scopes": "",
            "x-content-type-options": "nosniff",
            "x-frame-options": "deny",
            "x-github-media-type": "github.v3; format=json",
            "x-github-request-id": "8C9A:5A0B:3870ED0:6EBC003:62BB2ACE",
            "x-oauth-scopes": "notifications, public_repo, read:discussion, read:enterprise, read:gpg_key, read:org, read:packages, read:public_key, read:repo_hook, read:user, repo:status, repo_deployment, user:email",
            "x-ratelimit-limit": "5000",
            "x-ratelimit-remaining": "4874",
            "x-ratelimit-reset": "1656433380",
            "x-ratelimit-resource": "core",
            "x-ratelimit-used": "126",
            "x-xss-protection": "0"
        },
        "data": {
            "message": "Not Found",
            "documentation_url": "https://docs.github.com/rest/reference/git#create-a-tree"
        }
    },
    "request": {
        "method": "POST",
        "url": "https://api.github.com/repos/wolfy1339/test-repository/git/trees",
        "headers": {
            "accept": "application/vnd.github.v3+json",
            "user-agent": "octoherd-cli/3.4.8 octoherd-cli/2.3.1 octokit-core.js/3.6.0 Node.js/16.15.0 (linux; x64)",
            "authorization": "token [REDACTED]",
            "content-type": "application/json; charset=utf-8"
        },
        "body": "{\"base_tree\":\"ea0e827bafdec61e54000a8631941f383f6dbdc6\",\"tree\":[{\"path\":\".github/workflows/test.yml\",\"mode\":\"100644\",\"content\":\"name: Test\\n\\\"on\\\":\\n  push:\\n    branches:\\n      - master\\n  pull_request:\\n    types:\\n      - opened\\n      - synchronize\\njobs:\\n  test_matrix:\\n    runs-on: ubuntu-latest\\n    strategy:\\n      matrix:\\n        node_version:\\n          - 14\\n          - 16\\n          - 18\\n    steps:\\n      - uses: actions/checkout@v3\\n      - name: \\\"Use Node.js ${{ matrix.node_version }}\\\"\\n        uses: actions/setup-node@v3\\n        with:\\n          node-version: \\\"${{ matrix.node_version }}\\\"\\n          cache: npm\\n      - run: npm ci\\n      - run: npm test\\n\\n  test:\\n    runs-on: ubuntu-latest\\n    needs: test_matrix\\n    steps:\\n      - uses: actions/checkout@v3\\n      - run: npm ci\\n      - run: npm run test:typescript\\n      - run: npm run lint\\n\"}]}",
        "request": {}
    },
    "level": "error",
    "time": 1656433358280
}
```

wolfy1339 committed:
build(deps): update `octokit-plugin-create-pull-request`

wolfy1339 committed:
fix: allow to update existing PRs

oscard0m commented on script.js:
```diff
@@ -0,0 +1,186 @@
+        }
+    );
+
+    // We know that the path is a directory, we do this check to appease typescript and get rid of the `undefined`
+    if (Array.isArray(data.entries)) {
+      files = data.entries.filter(entry => isYamlFile(entry.name));
+    } else {
+      files = [data];
+    }
```
From what I see in `PATH` variable, it is a file and not a directory, right?
Should we check all the `yaml` files under `.github/workflows` just in case?

wolfy1339 replied:
>From what I see in PATH variable, it is a file and not a directory, right?
> Should we check all the yaml files under .github/workflows just in case?

That might actually be why i'm getting the 404's, I pointed it to a file and not the directory

oscard0m commented on script.js:
```diff
@@ -0,0 +1,186 @@
+  // Pull Request
+  //
+
+  const pr = await composeCreatePullRequest(octokit, {
+    owner,
+    repo,
+    title: `ci: stop testing against NodeJS ${NODE_VERSIONS_STRING}`,
+    body: `BREAKING CHANGE: Drop support for NodeJS ${NODE_VERSIONS_STRING}`,
+    head: branchName,
+    changes,
+    createWhenEmpty: false,
+    update: true
+  });
+
+
+  if (pr)
+    octokit.log.info(`Pull request created: ${pr.data.html_url}`);
```
Should we wrap this in a try-catch and `ocotkit.log.error` accordingly or do we want to just throw?

wolfy1339 replied:
I don't think it's warranted

oscard0m commented:
> I only seem to have one problem with the script.
> 
> The endpoint `/repos/{owner}/{repository}/git/trees` returns a 404 when trying to do a `POST` request (via `octokit-plugin-create-pull-request`)
> 
> Here's an extract of the log for the failed request
> 
<details>
  <summary>Click to expand!</summary>
  
  ```javascript
    {
    "repository": {
        "id": 366884413,
        "owner": {
            "login": "wolfy1339",
            "id": 4595477,
            "node_id": "MDQ6VXNlcjQ1OTU0Nzc=",
            "avatar_url": "https://avatars.githubusercontent.com/u/4595477?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/wolfy1339",
            "html_url": "https://github.com/wolfy1339",
            "followers_url": "https://api.github.com/users/wolfy1339/followers",
            "following_url": "https://api.github.com/users/wolfy1339/following{/other_user}",
            "gists_url": "https://api.github.com/users/wolfy1339/gists{/gist_id}",
            "starred_url": "https://api.github.com/users/wolfy1339/starred{/owner}{/repo}",
            "subscriptions_url": "https://api.github.com/users/wolfy1339/subscriptions",
            "organizations_url": "https://api.github.com/users/wolfy1339/orgs",
            "repos_url": "https://api.github.com/users/wolfy1339/repos",
            "events_url": "https://api.github.com/users/wolfy1339/events{/privacy}",
            "received_events_url": "https://api.github.com/users/wolfy1339/received_events",
            "type": "User",
            "site_admin": false
        },
        "name": "test-repository"
    },
    "msg": "HttpError: Not Found\n    at /home/user/Desktop/Programming/octoherd-remove-eol-node-versions/node_modules/@octokit/request/dist-node/index.js:86:21\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async Job.doExecute (/home/user/Desktop/Programming/octoherd-remove-eol-node-versions/node_modules/bottleneck/light.js:405:18)",
    "name": "HttpError",
    "status": 404,
    "response": {
        "url": "https://api.github.com/repos/wolfy1339/test-repository/git/trees",
        "status": 404,
        "headers": {
            "access-control-allow-origin": "*",
            "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset",
            "connection": "close",
            "content-encoding": "gzip",
            "content-security-policy": "default-src 'none'",
            "content-type": "application/json; charset=utf-8",
            "date": "Tue, 28 Jun 2022 16:22:38 GMT",
            "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
            "server": "GitHub.com",
            "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
            "transfer-encoding": "chunked",
            "vary": "Accept-Encoding, Accept, X-Requested-With",
            "x-accepted-oauth-scopes": "",
            "x-content-type-options": "nosniff",
            "x-frame-options": "deny",
            "x-github-media-type": "github.v3; format=json",
            "x-github-request-id": "8C9A:5A0B:3870ED0:6EBC003:62BB2ACE",
            "x-oauth-scopes": "notifications, public_repo, read:discussion, read:enterprise, read:gpg_key, read:org, read:packages, read:public_key, read:repo_hook, read:user, repo:status, repo_deployment, user:email",
            "x-ratelimit-limit": "5000",
            "x-ratelimit-remaining": "4874",
            "x-ratelimit-reset": "1656433380",
            "x-ratelimit-resource": "core",
            "x-ratelimit-used": "126",
            "x-xss-protection": "0"
        },
        "data": {
            "message": "Not Found",
            "documentation_url": "https://docs.github.com/rest/reference/git#create-a-tree"
        }
    },
    "request": {
        "method": "POST",
        "url": "https://api.github.com/repos/wolfy1339/test-repository/git/trees",
        "headers": {
            "accept": "application/vnd.github.v3+json",
            "user-agent": "octoherd-cli/3.4.8 octoherd-cli/2.3.1 octokit-core.js/3.6.0 Node.js/16.15.0 (linux; x64)",
            "authorization": "token [REDACTED]",
            "content-type": "application/json; charset=utf-8"
        },
        "body": "{\"base_tree\":\"ea0e827bafdec61e54000a8631941f383f6dbdc6\",\"tree\":[{\"path\":\".github/workflows/test.yml\",\"mode\":\"100644\",\"content\":\"name: Test\\n\\\"on\\\":\\n  push:\\n    branches:\\n      - master\\n  pull_request:\\n    types:\\n      - opened\\n      - synchronize\\njobs:\\n  test_matrix:\\n    runs-on: ubuntu-latest\\n    strategy:\\n      matrix:\\n        node_version:\\n          - 14\\n          - 16\\n          - 18\\n    steps:\\n      - uses: actions/checkout@v3\\n      - name: \\\"Use Node.js ${{ matrix.node_version }}\\\"\\n        uses: actions/setup-node@v3\\n        with:\\n          node-version: \\\"${{ matrix.node_version }}\\\"\\n          cache: npm\\n      - run: npm ci\\n      - run: npm test\\n\\n  test:\\n    runs-on: ubuntu-latest\\n    needs: test_matrix\\n    steps:\\n      - uses: actions/checkout@v3\\n      - run: npm ci\\n      - run: npm run test:typescript\\n      - run: npm run lint\\n\"}]}",
        "request": {}
    },
    "level": "error",
    "time": 1656433358280
}
  ```
</details>

Did you find out what was the issue with this? Do you need any help on debugging/reproducing this? @wolfy1339

wolfy1339 committed:
fix: don't look at a particular file, look at the directory

wolfy1339 committed:
fix: convert content to text even if the file isn't modified

wolfy1339 commented:
The 404s are gone now!

There is another case of outdated node versions where it isn't in a matrix, however those can be manually fixed

wolfy1339 commented:
Here's an example of what the output is: https://github.com/wolfy1339/test-repository/pull/10

oscard0m commented:
> Here's an example of what the output is: [wolfy1339/test-repository#10](https://github.com/wolfy1339/test-repository/pull/10)


LGTM!
